flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -216,13 +204,74 @@ public StreamGraph generate() {
 		return builtStreamGraph;
 	}
 
+	@VisibleForTesting
+	RuntimeExecutionMode determineExecutionMode(final RuntimeExecutionMode chosenMode) {
+		if (checkNotNull(chosenMode) != RuntimeExecutionMode.AUTOMATIC) {
+			return chosenMode;
+		}
+
+		final boolean continuousSourceExists = transformations
+				.stream()
+				.anyMatch(transformation ->
+						isContinuousSource(transformation) ||
+						transformation
+								.getTransitivePredecessors()
+								.stream()
+								.anyMatch(this::isContinuousSource));
+
+		return continuousSourceExists
+				? RuntimeExecutionMode.STREAMING
+				: RuntimeExecutionMode.BATCH;
+	}
+
+	private boolean isContinuousSource(final Transformation<?> transformation) {
```

Comment:
We try and keep the terminology consistent. Maybe this should be `isUnboundedSource()`?